domaincommon/src/state.rs

```diff
@@ -216,6 +221,26 @@ impl MatchState {
             .collect::<HashMap<_, _>>();
         state
     }
+
+    fn update_register(
+        state: &MatchState,
+        mech: &str,
+        reg: &u32,
+        val: &RegisterValue,
+    ) -> MatchState {
+        let mut state = state.clone();
```

Comment:
While I'm a fan of the functional style of building a new thing and returning, the state here is mutable... so there's nothing preventing us from doing something like `state.mechs.entry("mech").and_modify(|m| m.registers.insert(reg, val)`

Reply:
Just make sure to pepper it with the appropriate `or_insert`s so that you can receive a register update for a mech that might not (yet) exist.

Reply:
Trying the following:
```
        state.mechs.entry(mech.to_string()).and_modify(|m| {
            m.registers.insert(*reg, val.clone());
            ()
        });
        state
```

results in the error

```
cannot borrow `state.mechs` as mutable, as it is behind a `&` reference
`state` is a `&` reference, so the data it refers to cannot be borrowed as mutablerustc(E0596)
state.rs(232, 9): `state` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```

What do you think about the following snippet instead of cloning a few things, just following the previous pattern of cloning state?
```
        let mut state = state.clone();
        state.mechs.entry(mech.to_string()).and_modify(|m| {
            m.registers.insert(*reg, val.clone());
            ()
        });
        state
```

Reply:
Looks good